pyro/ops/streaming.py

```diff
@@ -0,0 +1,277 @@
+        >>> a_stats.update(torch.tensor(0.))
+        >>> b_stats.update(torch.tensor([1., 2.]))
+        >>> summary = {"a": a_stats.get(), "b": b_stats.get()}
+
+        # Version 2. Collect samples into dictionaries.
+        >>> stats = StatsOfDict({"a": CountStats, "b": CountMeanStats})
+        >>> stats.update({"a": torch.tensor(0.), "b": torch.tensor([1., 2.])})
+        >>> summary = stats.get()
+        >>> summary
+        {'a': {'count': 1}, 'b': {'count': 1, 'mean': tensor([1., 2.])}}
+
+    :param default: Default type of statistics of values of the dictionary.
+        Defaults to the inexpensive :class:`CountStats`.
+    :param dict types: Dictionary mapping key to type of statistic that should
+        be recorded for values corresponding to that key.
+    """
+
+    def __init__(
+        self,
+        types: Dict[Hashable, Callable[[], StreamingStats]] = {},
+        default: Callable[[], StreamingStats] = CountStats,
+    ):
+        self.stats: Dict[Hashable, StreamingStats] = defaultdict(default)
+        self.stats.update({k: v() for k, v in types.items()})
+        super().__init__()
+
+    def update(self, sample: Dict[Hashable, Any]) -> None:
+        for k, v in sample.items():
+            self.stats[k].update(v)
+
+    def merge(self, other: "StatsOfDict") -> "StatsOfDict":
+        assert isinstance(other, type(self))
+        result = copy.deepcopy(self)
+        for k in set(self.stats).union(other.stats):
+            if k not in self.stats:
+                result.stats[k] = copy.deepcopy(other.stats[k])
+            elif k in other.stats:
+                result.stats[k] = self.stats[k].merge(other.stats[k])
+        return result
+
+    def get(self) -> Dict[Hashable, Any]:
+        """
+        :returns: A dictionary of statistics. The keys of this dictionary are
+            the same as the keys of the samples from which this object is
+            updated.
+        :rtype: dict
+        """
+        return {k: v.get() for k, v in self.stats.items()}
+
+
+class StackStats(StreamingStats):
+    """
+    Statistic collecting a stream of tensors into a single stacked tensor.
+    """
+
+    def __init__(self):
+        self.samples = []
+
+    def update(self, sample: torch.Tensor) -> None:
+        assert isinstance(sample, torch.Tensor)
+        self.samples.append(sample)
+
+    def merge(self, other: "StackStats") -> "StackStats":
+        assert isinstance(other, type(self))
+        result = StackStats()
+        result.samples = self.samples + other.samples
+        return result
+
+    def get(self) -> Dict[str, Union[int, torch.Tensor]]:
+        """
+        :returns: A dictionary with keys ``count: int`` and (if any samples
+            have been collected) ``samples: torch.Tensor``.
+        :rtype: dict
+        """
+        if not self.samples:
+            return {"count": 0}
+        return {"count": len(self.samples), "samples": torch.stack(self.samples)}
```

Comment:
Could `"samples": ...` return a dict so we can do `sample_mean = samples['y'].mean()` in tests?